src/cargo/core/resolver/dep_cache.rs

```diff
@@ -513,25 +515,60 @@ impl RequirementError {
                     .filter(|dep| dep.name_in_toml() == feat)
                     .collect();
                 if deps.is_empty() {
+                    let closest =
+                        closest_msg(&feat.as_str(), summary.features().keys(), |key| &key);
                     return match parent {
                         None => ActivateError::Fatal(anyhow::format_err!(
-                            "Package `{}` does not have the feature `{}`",
+                            "Package `{}` does not have the feature `{}`{}",
                             summary.package_id(),
-                            feat
+                            feat,
+                            closest
                         )),
                         Some(p) => {
                             ActivateError::Conflict(p, ConflictReason::MissingFeatures(feat))
                         }
                     };
                 }
                 if deps.iter().any(|dep| dep.is_optional()) {
+                    let mut features = vec![];
+                    for (name, values) in summary.features() {
+                        for value in values {
+                            match value {
+                                FeatureValue::Dep { dep_name }
+                                | FeatureValue::DepFeature {
+                                    dep_name,
+                                    weak: false,
+                                    ..
+                                } if dep_name == &feat => {
+                                    features.push(*name);
+                                    // Don't add this feature multiple times
+                                    break;
+                                }
+                                _ => (),
+                            }
+                        }
+                    }
+                    let mut suggestion = String::new();
+                    if !features.is_empty() {
+                        // Features is already sorted because it was constructed from a BTreeMap.
+                        suggestion =
+                            String::from("\nThis feature would be enabled by these features:");
+                        let mut features = features.iter();
+                        for feature in (&mut features).take(3) {
+                            let _ = write!(&mut suggestion, "\n\t- `{}`", feature);
+                        }
+                        if features.next().is_some() {
+                            suggestion.push_str("\n\t  ...");
+                        }
+                    }
```

Comment:
1. Is there a reason this isn't being done in the only relevant branch?
2. Should we move some or all of this out to a helper function to not detract from the flow of this code?a